src/coreclr/jit/morph.cpp

```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
     return op1;
 }
 
+//------------------------------------------------------------------------------
+// fgMorphOpCast : Try to simplify "(T)x op (T)y" to "(T)(x op y)".
+// https://github.com/dotnet/runtime/issues/13816
+//
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
+//
+
+GenTree* Compiler::fgMorphOpCast(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_OR, GT_AND, GT_XOR));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+    genTreeOps oper = tree->OperGet();
+
+    // see whether both ops are casts, with matching to and from types
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
```

Comment:
both op2 (or at least op1?) have to be also checked for IsOverflow, because `(T)op1` might cause an OverflowException

Reply:
cdec019d463276d2546e47e51d6894e951f0a7b7